### ISSUE TITLE
callWhenRunning is calling the function before the signal handlers are installed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3146]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3146|
|Type|release blocker: regression|
|Created|2008-04-01 12:25:39Z|
|Branch|https://github.com/twisted/twisted/tree/trivial-sysev-3146-2|

When we spawn a process in a callWhenRunning method, we can see the warning: "spawnProcess called, but the SIGCHLD handler is not installed. ...". It seems to work well, but the message is wrong.

The root of this problem is that we used to do a callLater between startup events, and don't anymore. Maybe we can just move the initialization of signals handler sooner though.

<details><summary>Searchable metadata</summary>

```
trac-id__3146 3146
type__release_blocker__regression release blocker: regression
reporter__therve therve
priority__highest highest
milestone__ 
branch__branches_trivial_sysev_3146_2 branches/trivial-sysev-3146-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1207052739000000 1207052739000000
changetime__1333119747000000 1333119747000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve set owner to @exarkun|
|-|-|

What do you think? Should we add the callLater back?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

This is a dumb question (because I should write a unit test and use that to know the answer, but anyway), would this be a reasonable fix?

```
Index: twisted/internet/posixbase.py
===================================================================
--- twisted/internet/posixbase.py       (revision 23147)
+++ twisted/internet/posixbase.py       (working copy)
@@ -214,9 +214,9 @@
         """
         Forward call to ReactorBase, install signal handlers if asked.
         """
-        ReactorBase.startRunning(self)
         if installSignalHandlers:
             self._handleSignals()
+        ReactorBase.startRunning(self)
 
     def run(self, installSignalHandlers=True):
         self.startRunning(installSignalHandlers=installSignalHandlers)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

Well, that fixes the reported problem, sure. But the fact is that the 'after startup' event is no longer ran once the reactor has started. I'm not sure if it can't have other side effects.

Also, we have to check for every reactors.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Ah, that's a good point.  I think we should clarify what "after startup" is actually supposed to mean.  This is both a documentation and a testing issue, I guess.  If clarification shows that we should put the `callLater` back, then I won't be too opposed to it (but I'll be happier if it turns out to not be necessary).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

[#3168](https://github.com/twisted/twisted/issues/3168) is another manifestation of this change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [23234]) Branching to 'system-event-ordering-3146'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [23280]) Branching to 'trivial-sysev-3146'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [23295]) Branching to 'trivial-sysev-3146-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

This issue should be fixed in the branch.  `system-event-ordering-3146` got bigger than I wanted, so I pulled out the unit tests that were directly applicable to this ticket and to [#3168](https://github.com/twisted/twisted/issues/3168), dropped all the fancy test framework stuff, and then made the simplest change I could to the implementation to make them pass.

So, the `trivial-sysev-3146-2` should resolve this ticket and [#3168](https://github.com/twisted/twisted/issues/3168).

I'll make a new ticket later for the test improvements in the old branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Not a full review, but some comments: 
 * trial starts printing a warning when running process. It was already there but in the log file. I suspect this could be disturbing. Maybe we could explicitely call _handleSignals? I don't have a better (or worse) suggestion right now.
 * I don't like the addition of a new flag :/

Otherwise, it's very well commented and documented, and the buildbot is green. Maybe I ask to much :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

I filed [#3178](https://github.com/twisted/twisted/issues/3178) for the first of these.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @exarkun|
|-|-|

OK, a full review:
  * The test case should be more selective about the processes it tears down; it should just wait for its own processes to exit, not shotgun through all reaping.  I wish I could say something about the signal handling being more selective too, but I think it's doing as well as it can given the limitations of the platform.
  * The test case should have a more descriptive docstring; I don't think there was any test coverage for this functionality independent of the regressions, so it should just describe the properties it is verifying.  Normally I say "this is a regression test" if I am verifying something I wouldn't normally, such as the absence of some behavior, but these tests look pretty decent otherwise.
  * Make a new ticket for the contents of the system-event-ordering branch, and refer to that from the comment about `.clock.advance(0)`, so that users know what a better approach might be.

Modulo these minor changes, I believe this is ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [23340]) use os.waitpid instead of reapAllProcesses

refs [#3146](https://github.com/twisted/twisted/issues/3146)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [23341]) slightly better docstring for SystemEventOrderRegressionTests

refs [#3146](https://github.com/twisted/twisted/issues/3146)
;

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [23342]) Refer to another ticket

refs [#3168](https://github.com/twisted/twisted/issues/3168)
refs [#3146](https://github.com/twisted/twisted/issues/3146)
refs [#3198](https://github.com/twisted/twisted/issues/3198)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [23344]) put back SIG_DFL before calling waitpid

refs [#3146](https://github.com/twisted/twisted/issues/3146)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [23345]) Merge trivial-sysev-3146-2

Author: exarkun, radix
Reviewer: glyph
Fixes: [#3146](https://github.com/twisted/twisted/issues/3146)

Change reactor startup so that calls scheduled with callWhenRunning are only
processed after the reactor is properly started up.

Change reactor shutdown so that calling reactor.stop inside an event handler
(such as clientConnectionFailed) is safe again.

Change the warning emitted by the process support code to use the warnings
module instead of normal Twisted logging.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
